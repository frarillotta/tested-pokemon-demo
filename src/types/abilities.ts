import type { URLString } from './pokemon';

export type Abilities = {
    effect_changes: [],
    effect_entries: {
          effect: string,
          language:{
             name: string,
             url:URLString
          },
          short_effect:string
    }[],
    flavor_text_entries: {
          flavor_text: string,
          language:{
             name: string,
             url: URLString
          },
          version_group:{
            name: string,
            url: URLString
          }
    }[],
    generation:{
       name: string,
       url: URLString
    },
    id: number,
    is_main_series: boolean,
    name: string,
    names:{
          language:{
            name: string,
            url: URLString
          },
          name: string
       }[],
    pokemon: {
          is_hidden: boolean,
          pokemon:{
            name: string,
            url: URLString
          },
          slot: number
    }[]
 }