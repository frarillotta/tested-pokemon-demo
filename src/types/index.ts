export type { pokemonType, pokemonResults } from './pokemonsList'; 
export type { Pokemon, GenerationSprites, URLString, Stat, Move } from "./pokemon"