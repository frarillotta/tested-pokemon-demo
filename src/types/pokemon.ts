import { GenerationSprites } from './pokemon';
export type Pokemon = {
    abilities: {
          ability:{
             name: string,
             url: URLString
          },
          is_hidden: boolean,
          slot: number
       }[],
    base_experience: number,
    forms: {
          name: string,
          url:URLString
       }[],
    game_indices: {
          game_index: number,
          version:{
             name: string,
             url: URLString
          }
    }[],
    height: number,
    held_items: any[],
    id: number,
    is_default: boolean,
    location_area_encounters: string,
    moves: Move[],
    name: string,
    order: number,
    past_types: [],
    species:{
       name: string,
       url: string
    },
    sprites:{
		back_default: URLString | null,
		back_female: URLString | null,
		back_shiny: URLString | null,
		back_shiny_female: URLString | null,
		front_default: URLString | null,
		front_female: URLString | null,
		front_shiny: URLString | null,
		front_shiny_female: URLString | null,
       other:{
          dream_world: GenerationSprites,
          "official-artwork": GenerationSprites
       },
       versions:{
          "generation-i": {
             "red-blue": GenerationSprites,
             yellow: GenerationSprites
          },
            "generation-ii": {
             crystal: GenerationSprites,
             gold: GenerationSprites,
             silver: GenerationSprites
          },
          "generation-iii":{
             emerald: GenerationSprites,
             "firered-leafgreen": GenerationSprites,
             "ruby-sapphire": GenerationSprites
          },
          "generation-iv":{
             "diamond-pearl": GenerationSprites,
             "heartgold-soulsilver": GenerationSprites,
             platinum: GenerationSprites
          },
          "generation-v":{
             "black-white":{
                animated: GenerationSprites,
                GenerationSprites
             }
          },
          "generation-vi":{
             "omegaruby-alphasapphire": GenerationSprites,
             "x-y": GenerationSprites
          },
          "generation-vii":{
             icons: GenerationSprites,
             "ultra-sun-ultra-moon": GenerationSprites
          },
          "generation-viii":{
             icons: GenerationSprites
          }
       }
    },
    stats: Stat[],
    types: {
          slot: number,
          type:{
             name: "normal" | "fire" | "water" | "grass" | "electric" | "ice" | "fighting" | "poison" | "ground" | "flying" | "psychic" | "bug" | "rock" | "ghost" | "dark" | "dragon" | "steel" | "fairy",
             url: URLString
          }
       }[],
    weight: number
 }

export type GenerationSprites = {
    back_default: URLString | null,
    back_female: URLString | null,
    back_shiny: URLString | null,
    back_shiny_female: URLString | null,
    front_default: URLString | null,
    front_female: URLString | null,
    front_shiny: URLString | null,
    front_shiny_female: URLString | null
 };

export type Stat = {
	base_stat: number,
	effort: number,
	stat:{
	   name: "hp" | "attack" | "defense" | "special-defense" | "special-attack" | "speed",
	   url: URLString
	}
};

export type Move = {
	move:{
	   name: string,
	   url: URLString
	},
	version_group_details: {
		  level_learned_at: number,
		  move_learn_method:{
			 name: string,
			 url: URLString
		  },
		  version_group:{
			 name: string,
			 url: URLString
		  }
	  }[],
 }

export type URLString = `https://${string}`;
