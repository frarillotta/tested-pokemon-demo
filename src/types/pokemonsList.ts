export type pokemonType = {
    count: number,
    next: string,
    previous: string,
    results: pokemonResults[]
}

export type pokemonResults = {
    name: string, 
    url: string
}