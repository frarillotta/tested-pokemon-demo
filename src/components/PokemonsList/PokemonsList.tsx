import { pokemonResults } from '../../types/pokemonsList';
import { memo } from 'react';
import { FixedSizeList as List } from 'react-window';
import styled from "styled-components";
import { PokeballButton } from '../PokeballButton/PokeballButton';
import AutoSizer from "react-virtualized-auto-sizer";
import { capitalizeFirstLetter } from "../../utils";

type PokemonsListProps = {
	className: string,
    pokemons: pokemonResults[],
    capturedPokemons: Record<string, string>
    removePokemon: (pokemonName: string) => void;
    addPokemon: (pokemonResults: pokemonResults) => void;
	setActivePokemon: (pokemon: pokemonResults) => void;
}

const PokemonsList = memo(({className, pokemons, capturedPokemons, removePokemon, addPokemon, setActivePokemon}: PokemonsListProps) => {

	const Row = ({ index, style }) => {
		const pokemon = pokemons[index];
		const pokemonName = pokemon && capitalizeFirstLetter(pokemon.name);
		const isCaptured = pokemon && capturedPokemons.hasOwnProperty(pokemon.name);
		return <>
			<ListItem 
				onClick={()=> setActivePokemon(pokemon)}
				index={index}
				style={style}
			>
				{pokemonName}
			</ListItem>
			<Pokeball 
				style={{ ...style, position: "absolute", left: "none", width: "none", right: "0px" }}
				isCaptured={isCaptured} 
				onClick={() => isCaptured ? removePokemon(pokemon.name) : addPokemon(pokemons[index])}
				size={30}
			/>
		</>
	};

	return (
		<Wrapper className={className}>
			<AutoSizer>
				{({ height, width }) => (
					<List
						height={height}
						itemCount={pokemons.length}
						itemSize={35}
						width={width}
					>
						{Row}
					</List>
				)}
			</AutoSizer>
		</Wrapper>
	)
});
PokemonsList.displayName = "PokemonsList"

const Wrapper = styled.main`
	width: 100%;
	height: 100%;
`

const Pokeball = styled(PokeballButton)`
	display: inline-flex;
	align-items: center;
	padding-right: 10px;
`

const ListItem = styled.span`
	cursor: pointer;
	display: inline-flex;
	padding-left: 10px;
	align-items: center;
	background: ${props => props.index % 2 === 0 ? "#ffb50021" : "none"};
	border-radius: 10px;
`


export { PokemonsList }