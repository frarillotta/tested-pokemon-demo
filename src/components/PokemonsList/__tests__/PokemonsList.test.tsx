import { render } from "@testing-library/react";
import {PokemonsList} from "../PokemonsList";
import list from "../../../../__mocks__/dataMocks/list.json";
import capturedPokemons from "../../../../__mocks__/dataMocks/capturedpokemon.json";
import { pokemonResults } from "../../../types";

describe("PokemonsList", () => {
	
	describe("when loaded", () => {
		
		it("should have a consistent snapshot", () => {
			const {container} = render(<PokemonsList 
				pokemons={list as unknown as pokemonResults[]} 
				addPokemon={jest.fn()}
				capturedPokemons={capturedPokemons}
				setActivePokemon={jest.fn()}
				removePokemon={jest.fn()}
			/>);
			expect(container).toMatchSnapshot();
		});

	});

})