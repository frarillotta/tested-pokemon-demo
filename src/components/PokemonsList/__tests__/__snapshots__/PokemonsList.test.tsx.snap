// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeballButton when isCaptured=true should have a consistent snapshot 1`] = `
<div>
  <main
    class="PokemonsList__Wrapper-sc-1vm6xrc-0 gEFRsL"
  >
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    />
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1px; height: 1px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </main>
</div>
`;

exports[`PokemonsList when loaded should have a consistent snapshot 1`] = `
<div>
  <main
    class="PokemonsList__Wrapper-sc-1vm6xrc-0 gEFRsL"
    id="pokemon-list"
  >
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    />
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1px; height: 1px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </main>
</div>
`;
